OpenRA.Game/Activities/Activity.cs

```diff
@@ -101,8 +101,13 @@ internal void OnActorDisposeOuter(Actor self)
 
 		public virtual void Cancel(Actor self, bool keepQueue = false)
 		{
-			if (!keepQueue)
+			if (!keepQueue && NextActivity != null)
+			{
+				foreach (var a in NextActivity.ActivitiesImplementing<IActivityNotifyPreemptiveCancel>())
```

Comment:
This kind of goes against the guidelines for how activities should behave that we hashed out in #16069 and other places. It is also only really relevant for `AttackFollow` and related activities and the bug this is meant to fix is caused specifically by `AttackFollow.OnQueueAttackActivity` and the resulting ambiguity over whether it is the trait or the activity that is in charge of targeting.

A proper solution would be to completely separate targeting from the activities. That would require a bigger refactor though, so doing that might better be deferred to the future.

In the mean time, if you think this absolutely needs to be fixed now, I would prefer it if this was kept contained to the activities for which it is relevant by using an override for `Cancel` instead of adding this to the base `Activity` class. the planned follow-ups to #16696 will merge these activities further so will likely negate the code duplication this might otherwise cause.

Reply:
The issue is that Cancel is not called for queued activities, so there is nothing to override.

It should be possible now to reimplement this from the trait tick, so I will try that instead.